docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+means of a function pointer table, where the position is the index in the code
+generation stage above, and the buffer is the contents of the arguments and a
+place for the return value.
+
+Argument decomposition will be done by the sandbox code that will call the actual
+function with the actual arguments and, if there is a return value, set it on
+the right place of the buffer and return.
+
+### Memory Management
+
+To avoid the sandbox from requesting memory in the wrong places (accidentally
+or maliciously), we create a two-tier allocator. The actual allocator in the
+Verona side, and a dummy allocator on the sandbox side that redirects all its
+calls to the Verona allocator. This is part of the sandbox library's
+functionality, which is passed along on the creation of the sandbox.
+
+The Verona allocator creates a heap for the sandbox and that's the memory that
+the sandbox code can use for everything. The sandbox code can also access that
+memory and it uses to keep the arguments and return values, so that the sandbox
+can see them.
+
+Because that heap is in a Verona sandbox region, it is safe to assume there are
+no race conditions introduced by the compiler onto the external library, due to
+the use of cowns to execute the foreign code.
+
+## Step-by-step example
+
+This is the description of a mock implementation of the child process sandbox
+running the C++ snippet in the beginning of this document. We'll go line by line
+and describe what the compiler and the generated code will do for each of them.
+
+### Compile time
+
+```c
+type StdVec = "c++::vector";
+```
+
+* Call the C++ driver (clang wrapper) to parse the include file (and all its
+  includes) and keep as a C++ AST.
+* Create a Verona class called `StdVec` with C++ built-in types defined as Verona
+  types (ex. `int` as `S32`, `size_t` as `U64`).
+* Expose an interface for the parser to query symbols (types, objects and
+  functions).
+
+```c
+reg = new StdVec;
+```
+
+* The parser recognises the region as a C++ sandbox.
+* Lower code to create a region with a sandbox allocator (RT calls).
+* Lower code to create the sandbox memory area (RT allocator calls).
+* Create a wrapper object for function dispatch and general sandbox utilities.
+* The region is bound to the variable `reg`.
+
+```c
+obj = new std::vector[int] in reg
+```
+
+* The parser recognises the region and tries to create a new object in the
+  sandbox region by using the C++ driver.
+* Assemble the foreign call signature (namespace, template args):
+  `std::vector<int> std::vector<int>::vector<int>()`
+* Instantiate the signature as a template specialisation in the AST, using
+  the query interface to validate the code and get an actual implementation.
+* If available, create the Verona equivalent inside the `StdVec` class and
+  Lower call to that function. Otherwise emit an error and stop processing.
+* The Verona signature would be something like
+  `create() : StdVec::std::vector[S32]`.
```

Comment:
I don't think it would be `S32`, it would be something like `ProcessSandbox::int` or `WAsmSandbox::int`.

Reply:
I just mention `S32` to give an idea what kind of Verona type it will map to. Of course it won't always be a signed 32-bit integer.

I don't think we can have a single `ProcessSandbox::int` for all targets as it depends on the target architecture and the foreign language standard rules.

Reply:
Possibly `ProcessSandbox::C::int`, but this is part of the sandbox-environment ABI, which is not necessarily the same for all instantiations of a foreign library.  We should be able (eventually) on the same system to instantiate:

 - An SFI sandbox, possibly via WebAssembly
 - A process-based sandbox
 - A CHERI CheriABI sandbox
 - A CHERI A64-ABI sandbox

On Morello, we'd end up with:

 - `WAsmSandbox::C::intptr_t` is a 32-bit integer, which we can losslessly round-trip to and from a `S32`.
 - `ProcessSandbox::C:intptr_t` is a 64-bit integer, which we can losslessly round-trip to and from `S64`.
 - `CheriABISandbox::C::intptr_t` is a capability (and not something we currently have a Verona type for - something we need only for C interop).
 - `CheriLegacySandbox::C::intptr_t` is a 64-bit integer, which we can losslessly round-trip to and from `S64`.

Ideally, we don't want these to be properties of the foreign module, because the same module should be a generic class that takes that is instantiated with one of any of these sandboxing models (or some per-module subset if we don't have a version of the library for all of them).

Reply:
I agree with everything, I just think this is too low level to be in this document. I'll just change `S32` to `Sandbox::int` and mention that it's sandbox specific.